include/ppx/metrics.h

```diff
@@ -0,0 +1,198 @@
+// limitations under the License.
+
+#ifndef ppx_metrics_h
+#define ppx_metrics_h
+
+#include "ppx/config.h"
+
+#include <limits>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace ppx {
+namespace metrics {
+
+#define METRICS_NO_COPY(type)   \
+    type(type&)       = delete; \
+    type(const type&) = delete; \
+    type& operator=(const type&) = delete;
+
+////////////////////////////////////////////////////////////////////////////////
+
+enum class MetricInterpretation
+{
+    NONE,
+    HIGHER_IS_BETTER,
+    LOWER_IS_BETTER,
+};
+
+struct Range
+{
+    double lowerBound = std::numeric_limits<double>::min();
+    double upperBound = std::numeric_limits<double>::max();
+};
+
+struct MetricMetadata
+{
+    std::string          name;
+    std::string          unit;
+    MetricInterpretation interpretation = MetricInterpretation::NONE;
+    Range                expectedRange;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+
+enum class MetricType
+{
+    GAUGE,
+    COUNTER,
+};
+
+class Metric
+{
+    friend class Run;
+
+public:
+    const MetricMetadata& GetMetadata() const;
+    MetricType            GetType() const;
+
+protected:
+    Metric(const MetricMetadata& metadata, MetricType type);
+    virtual ~Metric();
+
+private:
+    MetricMetadata mMetadata;
+    MetricType     mType;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+
+struct GaugeStatistics
+{
+    double min;
+    double max;
+    double average;
+    double median;
+    double standardDeviation;
+    double timeRatio;
+    double percentile90;
+    double percentile95;
+    double percentile99;
+};
+
+class MetricGauge final : public Metric
+{
+public:
+    MetricGauge(const MetricMetadata& metadata);
+
+    void   RecordEntry(double seconds, double value);
+    size_t GetEntriesCount() const;
+    void   GetEntry(size_t index, double& seconds, double& value) const;
+
+    const GaugeStatistics GetStatistics(bool realtime) const;
+
+private:
+    struct TimeSeriesEntry
+    {
+        double seconds;
+        double value;
+    };
+
+private:
+    ~MetricGauge();
+    METRICS_NO_COPY(MetricGauge)
+
+    void UpdateRealTimeStatistics(double seconds, double value);
+
+private:
+    std::vector<TimeSeriesEntry> mTimeSeries;
+    GaugeStatistics              mRealTimeStatistics;
+    double                       mAccumlatedValue;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+
+class MetricCounter final : public Metric
+{
+    friend class Run;
+
+public:
+    MetricCounter(const MetricMetadata& metadata);
+    uint64_t Increment(uint64_t add);
+    uint64_t Get() const;
+
+private:
+    ~MetricCounter();
+    METRICS_NO_COPY(MetricCounter)
+
+private:
+    uint64_t mCounter;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+
+class Run final
```

Comment:
I'm guessing this Manager/Run/Metric hierarchy is for the following use case:
 - 1 benchmark -> 1 manager
 - this benchmark has 2 things to test: 2 runs
 - each run as N metrics to monitor: N metrics
 
 What bugs me is the following:

* runs are independent.
  They don't need to share anything. Can have a different set of metrics attached to it. Should be name be different? (run feels like doing multiple runs of the same benchmark). Maybe something more generic, like Collection/Group?
  If the goal is indeed to have the same metric set, shouldn't the manager value be to check that? 

* No concept of "live/archived" run:
 We run a first test with Run#1. Then run another test with Run#2. If my metrics pointer are mishandled, I can still modify metrics of Run#1, no way to enforce this. Meaning the only value Run/Manager brings is allocation handling. Could it be beneficial for the manager to mark metrics/runs are immutable, and assert if a metric is updated outside of a run? (Creating a new run marks the old one immutable by ex).

Reply:
In the context of the benchmark, the expected flow is:
1. set the knobs to particular values
2. create a run
3. add metrics to the run that are relevant to the knobs
4. record time series for each metrics
5. repeat for as long as needed

Yes, runs are independent. To be honest, I could not find a better name. I thought about benchmark, but that would be even more confusing for obvious reasons. Collection/Group feel a bit too generic. I will add a comment over the class to try and make it more obvious what it is.

It might definitely be beneficial to block editing... that should be easy. I will think about it in a subsequent PR, once I implement the rest of the system and start seeing more use cases.